zend_abstract_interface/hook/hook.c

```diff
@@ -43,31 +43,31 @@ typedef struct {
     size_t dynamic_offset;
 } zai_hook_info;
 
-ZEND_TLS zend_ulong zai_hook_invocation = 0;
-ZEND_TLS zend_ulong zai_hook_id;
-
 /* {{{ private tables */
+ZEND_TLS struct {
+    zend_ulong invocation;
+    zend_ulong id;
+    // zai_hook_tls->request_functions is a map name -> array<zai_hook_t>
+    HashTable request_functions;
+    // zai_hook_tls->request_classes is a map class name -> map function name -> array<zai_hook_t>
+    HashTable request_classes;
+    // zai_hook_tls->request_files is an array<zai_hook_t>
+    zai_hooks_entry request_files;
+    // zai_hook_tls->inheritors is a map of class entries (interfaces and abstract classes) to a list of class entries
+    HashTable inheritors;
+} *zai_hook_tls;
+
 // zai_hook_static is a simple array of persistently allocated zai_hook_t
 // these persistently allocated zai_hook_t are always duplicated (with is_global = true) into zai_hook_request_* on request start
 static HashTable zai_hook_static;
 
-// zai_hook_request_functions is a map name -> array<zai_hook_t>
-ZEND_TLS HashTable zai_hook_request_functions;
-// zai_hook_request_classes is a map class name -> map function name -> array<zai_hook_t>
-ZEND_TLS HashTable zai_hook_request_classes;
-// zai_hook_request_files is an array<zai_hook_t>
-ZEND_TLS zai_hooks_entry zai_hook_request_files;
-
 // zai_hook_resolved is a map op_array/internal_function -> array<zai_hook_t>
-// if indirect, then it's pointing to some hashtable in zai_hook_request_functions/classes
+// if indirect, then it's pointing to some hashtable in zai_hook_tls->request_functions/classes
 TSRM_TLS HashTable zai_hook_resolved;
 
-// zai_hook_inheritors is a map of persistent class entries (interfaces and abstract classes) to a list of persistent class entries
+// zai_hook_static_inheritors is a map of persistent class entries (interfaces and abstract classes) to a list of persistent class entries
 static HashTable zai_hook_static_inheritors;
 
-// zai_hook_inheritors is a map of class entries (interfaces and abstract classes) to a list of class entries
-ZEND_TLS HashTable zai_hook_inheritors;
-
```

Comment:
**question (non-blocking):** Just for me to recap what is going in, as my TLS-Fu is weak.
Those changes move the variables into a struct and out of the TLS. The thing remaining in TLS is a pointer to that struct and in GINIT you allocate this and deallocate in GSHUTDOWN, correct?

Reply:
yes :-)